packages/publish/src/publish.ts

```diff
@@ -105,6 +110,154 @@ function handler(app: Application) {
+      withFileTypes: true,
+    });
+    let pkgPath = unpackPath;
+    if (files.length === 1 && files[0].isDirectory()) {
+      pkgPath = `${pkgPath}/${files[0].name}`;
+    }
+
+    // Update the package.json on disk with a candidate version number, i.e.,
+    // v3.0.0-beta.0:
+    const candidateVersion = await setCandidateVersion(
+      pkgPath,
+      context.payload.pull_request.title
+    );
+
+    const secret: Secret = await handler.getPublicationSecrets(
+      app,
+      remoteConfiguration
+    );
+    if (secret && secret.payload && secret.payload.data) {
+      // Create a pre-release on GitHub for this release candidate:
+      try {
+        await context.github.repos.createRelease({
+          owner: context.payload.repository.owner.login,
+          repo: context.payload.repository.name,
+          tag_name: `v${candidateVersion.version}`,
+          name: `v${candidateVersion.version} Pre-release`,
+          prerelease: true,
+          target_commitish: 'master',
+          body: `This is a pre-release for testing purposes of #${context.payload.pull_request.number}.\n\nInstall by running \`npm install ${candidateVersion.name}@next\``,
+        });
+      } catch (err) {
+        // The most likely error is that the release already existed, and we
+        // had a publication failure; we can skip an error creating the release:
+        app.log.error(err);
+      }
+
+      // Publish the package:
+      const publishConfig = handler.publishConfigFromSecret(secret);
+      const npmRc = handler.generateNpmRc(publishConfig);
+      await handler.publish(npmRc, pkgPath, app, true);
+
+      await context.github.issues.removeLabel({
+        owner: context.payload.repository.owner.login,
+        repo: context.payload.repository.name,
+        issue_number: context.payload.pull_request.number,
+        name: PRE_RELEASE_LABEL,
+      });
+      await context.github.issues.createComment({
+        owner,
+        repo: repoName,
+        issue_number: context.payload.pull_request.number,
+        body: `A candidate release, \`${candidateVersion.version}\` was published to npm. Run \`npm install ${candidateVersion.name}@next\` to install.`,
+      });
+    } else {
+      app.log.error('could not load application secrets');
+    }
+  });
+}
+
+async function setCandidateVersion(
+  pkgPath: string,
+  title: string
+): Promise<CandidateVersionInfo> {
+  // eslint-disable-next-line @typescript-eslint/no-var-requires
+  const pkgJson = require(`${pkgPath}/package.json`);
+  const futureVersion = title.match(
+    /release (?<version>[0-9]+\.[0-9]+\.[0-9]+$)/
+  )?.groups?.version;
+  const versions = await fetch(
+    `https://registry.npmjs.org/${encodeURIComponent(pkgJson.name)}`
+  )
+    .then(res => {
+      return res.json();
+    })
+    .then(body => {
+      return Object.keys(body.time);
+    });
+  let counter = 0;
+  let candidateVersion = `${futureVersion}-beta.${counter}`;
```

Comment:
My brain melted trying to understand what's happening here.  Maybe a comment explaining?